web/modules/custom/search_api_field/src/Plugin/facets/facet_source/SearchApiField.php

```diff
@@ -0,0 +1,153 @@
+use Drupal\search_api\Entity\Index;
+use Drupal\search_api\Query\ResultSetInterface;
+
+/**
+ * Represents a facet source which represents search_api_page pages.
+ *
+ * Most of the work of actually getting a page is done in the deriver.
+ *
+ * @FacetsFacetSource(
+ *   id = "search_api_field",
+ *   deriver = "Drupal\search_api_field\Plugin\facets\facet_source\SearchApiFieldDeriver"
+ * )
+ */
+class SearchApiField extends SearchApiBaseFacetSource implements SearchApiFacetSourceInterface {
+
+  /**
+   * The entity manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManager|null
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The typed data manager.
+   *
+   * @var \Drupal\Core\TypedData\TypedDataManager|null
+   */
+  protected $typedDataManager;
+
+  /**
+   * The config factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface|null
+   */
+  protected $configFactory;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, array $plugin_definition, $query_type_plugin_manager, $search_results_cache) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $query_type_plugin_manager, $search_results_cache);
+
+    // Load facet plugin definition and depending on those settings; load the
+    // corresponding search api page and load its index.
+    $field_id = $plugin_definition['search_api_field'];
+    /* @var $page \Drupal\search_api\Utility\QueryHelper */
+    $field = FieldStorageConfig::load($field_id);
+    $index = $field->getSetting('index');
+    $this->index = Index::load($index);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function fillFacetsWithResults($facets) {
+    // Check if there are results in the static cache.
+    $results = $this->searchApiQueryHelper->getResults($this->pluginId);
+
+    // If there are no results, execute the search page and check for results
+    // again. This happens when a page or block is cached, so Search API has
+    // not fired an actual search.
+    if (!$results) {
+      /* @var $search_api_index \Drupal\search_api\IndexInterface */
+      $search_api_index = $this->getIndex();
+
+      // Create the query.
+      $options = [
+        'parse_mode' => 'direct',
+        // @Todo Fix limit, get it from field settings.
+        'limit' => 10,
+        'offset' => isset($_GET['page']) ? $_GET['page'] : 0,
+        'search id' => $this->pluginId,
+      ];
+      $query = $search_api_index->query($options);
+
+      // Keys.
+      $keys = \Drupal::request()->get('keys');
+      if (!empty($keys)) {
+        $query->keys($keys);
+      }
+
+      // Index fields.
+      $query->setFulltextFields();
+
+      // Execute the query.
+      $results = $query->execute();
+    }
+
+    // If we got results from the cache, this is the first code executed in this
+    // method, so it's good to double check that we can actually work with
+    // $results.
+    if ($results instanceof ResultSetInterface) {
```

Comment:
Ah! How great life would be if PHP would allow to specify return types! Hmmmmm, wait a minute...

:-1: for PHP5 compatibility :rage: